File: c/src/audio/wav.c
#include "ondes/wav.h"

#include <stdint.h>
#include <stdio.h>

#define BITS_PER_SAMPLE 16
#define NUM_CHANNELS 1

static void put_u32(FILE *f, uint32_t v) {
    uint8_t b[4] = {(uint8_t)v, (uint8_t)(v >> 8), (uint8_t)(v >> 16),
                    (uint8_t)(v >> 24)};
    fwrite(b, 1, 4, f);
}

static void put_u16(FILE *f, uint16_t v) {
    uint8_t b[2] = {(uint8_t)v, (uint8_t)(v >> 8)};
    fwrite(b, 1, 2, f);
}

int wav_write_mono16(const char *path, const double *samples, size_t n,
                     int sample_rate) {
    FILE *f = fopen(path, "wb");
    if (!f) return -1;

    uint32_t byte_rate = (uint32_t)sample_rate * NUM_CHANNELS * BITS_PER_SAMPLE / 8;
    uint16_t block_align = NUM_CHANNELS * BITS_PER_SAMPLE / 8;
    uint32_t data_size = (uint32_t)(n * NUM_CHANNELS * BITS_PER_SAMPLE / 8);

    fwrite("RIFF", 1, 4, f);
    put_u32(f, 36 + data_size);
    fwrite("WAVE", 1, 4, f);

    fwrite("fmt ", 1, 4, f);
    put_u32(f, 16);
    put_u16(f, 1); // PCM
    put_u16(f, NUM_CHANNELS);
    put_u32(f, (uint32_t)sample_rate);
    put_u32(f, byte_rate);
    put_u16(f, block_align);
    put_u16(f, BITS_PER_SAMPLE);

    fwrite("data", 1, 4, f);
    put_u32(f, data_size);

    for (size_t i = 0; i < n; i++) {
        int32_t v = (int32_t)(samples[i] * 32767);
        if (v > 32767)
            v = 32767;
        else if (v < -32768)
            v = -32768;
        put_u16(f, (uint16_t)(int16_t)v);
    }
    int ok = !ferror(f);
    fclose(f);
    return ok ? 0 : -1;
}

File: audio/wav.go
// Package audio provides WAV file I/O for the renderer. Mono, 16-bit
// PCM, 44.1 kHz - matching what regression/summarize.py expects.
package audio

import (
	"encoding/binary"
	"fmt"
	"os"
)

const (
	SampleRate    = 44100
	BitsPerSample = 16
	NumChannels   = 1
)

// WriteMono16 writes a mono 16-bit PCM WAV file at the given sample
// rate. Input is float64 samples in [-1, +1]; values outside that range
// are hard-clipped at the int16 boundary.
func WriteMono16(path string, samples []float64, sampleRate int) error {
	f, err := os.Create(path)
	if err != nil {
		return err
	}
	defer f.Close()

	byteRate := sampleRate * NumChannels * BitsPerSample / 8
	blockAlign := uint16(NumChannels * BitsPerSample / 8)
	dataSize := uint32(len(samples) * NumChannels * BitsPerSample / 8)

	// RIFF header
	if _, err := f.Write([]byte("RIFF")); err != nil {
		return err
	}
	if err := binary.Write(f, binary.LittleEndian, uint32(36+dataSize)); err != nil {
		return err
	}
	if _, err := f.Write([]byte("WAVE")); err != nil {
		return err
	}

	// fmt chunk
	if _, err := f.Write([]byte("fmt ")); err != nil {
		return err
	}
	if err := binary.Write(f, binary.LittleEndian, uint32(16)); err != nil {
		return err
	}
	if err := binary.Write(f, binary.LittleEndian, uint16(1)); err != nil { // PCM
		return err
	}
	if err := binary.Write(f, binary.LittleEndian, uint16(NumChannels)); err != nil {
		return err
	}
	if err := binary.Write(f, binary.LittleEndian, uint32(sampleRate)); err != nil {
		return err
	}
	if err := binary.Write(f, binary.LittleEndian, uint32(byteRate)); err != nil {
		return err
	}
	if err := binary.Write(f, binary.LittleEndian, blockAlign); err != nil {
		return err
	}
	if err := binary.Write(f, binary.LittleEndian, uint16(BitsPerSample)); err != nil {
		return err
	}

	// data chunk
	if _, err := f.Write([]byte("data")); err != nil {
		return err
	}
	if err := binary.Write(f, binary.LittleEndian, dataSize); err != nil {
		return err
	}

	buf := make([]byte, 2*len(samples))
	for i, s := range samples {
		v := int32(s * 32767)
		if v > 32767 {
			v = 32767
		} else if v < -32768 {
			v = -32768
		}
		binary.LittleEndian.PutUint16(buf[i*2:], uint16(int16(v)))
	}
	if _, err := f.Write(buf); err != nil {
		return fmt.Errorf("wav write: %w", err)
	}
	return nil
}
